docs/modules/third_party_rules_callbacks.md

```diff
@@ -148,6 +148,44 @@ deny an incoming event, see [`check_event_for_spam`](spam_checker_callbacks.md#c
 
 If multiple modules implement this callback, Synapse runs them all in order.
 
+### `on_profile_update`
+
+_First introduced in Synapse v1.5X.0_
+
+```python
+async def on_profile_update(
+    user_id: str,
+    new_profile: "synapse.module_api.ProfileInfo",
+    by_admin: bool,
+) -> None:
+```
+
+Called after updating a local user's profile. The update can be triggered either by the
+user themselves or a server admin. The update can also be triggered by a user being
+deactivated (in which case their display name is set to an empty string (`""`) and the
+avatar URL is set to `None`). The module is passed the Matrix ID of the user whose profile
+has been updated, their new profile, as well as a boolean that is `True` if the update
+was triggered by a server admin (and `False` otherwise). Note that this boolean is also
+`True` if the profile change happens as a result of the user logging through Single
+Sing-On, and if a server admin updates their own profile.
+
+If multiple modules implement this callback, Synapse runs them all in order.
+
+### `on_deactivation`
+
+_First introduced in Synapse v1.5X.0_
+
+```python
+async def on_deactivation(user_id: str, by_admin: bool) -> None:
```

Comment:
Note that we officially support re-activating users, which currently modules would not be told about: https://github.com/matrix-org/synapse/blob/18862f20b5495bdc556c54e92fd4b1efdc718ba7/synapse/rest/admin/users.py#L333-L335

It might be worthwhile to modify this callback to add support for information modules about re-activated users?

Reply:
I wonder if it's not better to add a `on_reactivation` callback in the future (when we end up needing one)? It feels more like the right way to convey this information to modules.

Reply:
Compiling both into one method would have the advantage of reducing the number of module callbacks we have, while potentially having the downside of a slightly more confusing function signature (`by_admin` would always be `True` on reactivation).

So I suppose it depends on how much value we put on keeping the callback surface small. If not much, then having both `on_deactivation` and `on_reactivation` methods are fine.

Small nit: `on_deactivation` -> `on_user_deactivation`